### PR TITLE
Add a simple Block Gossip routing

### PR DESCRIPTION
**Pull request by ybc1991**

### CODE REVIEW

File: scripts/mpi_passive_recv.cc [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#include <memory>
+#include <queue>
+#include <thread>
+#include <vector>
+
+const int kSendLength = 20;
+const int kTag = 1234567;
+
+#define MPICHECK(cmd)                                                  \
+  do {                                                                 \
```
Comment (hanbinhu) [inline scripts/mpi_passive_recv.cc:19]: Why use do while here? If it is for avoiding namespace contamination, I guess a pair of brackets should be enough.

File: bluefog/torch/mpi_ops.py [hunk 2]
```diff
@@ -463,6 +468,72 @@ def neighbor_allreduce_nonblocking(tensor: torch.Tensor,
...
+        _pair_gossip_nonblocking_function_factory, tensor)
+    handle = getattr(mpi_lib, function)(tensor, output, target_rank,
+                                        self_weight, pair_weight,
+                                        name.encode() if name is not None else "")
+    _handle_map[handle] = (tensor, output)
+    return handle
+
+
+# TODO(ybc) Consider active and passive version of pair gossip.
+def pair_gossip(tensor: torch.Tensor, target_rank: int, self_weight: float = 0.5,
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/torch/mpi_ops.py:486]: Ok. Make it as same style as neighbro allreduce.

File: bluefog/torch/mpi_ops.cc [hunk 3]
```diff
@@ -457,6 +457,74 @@ int DoNeighborAllreduce(::torch::Tensor tensor, ::torch::Tensor output,
...
+        bf_tensor, bf_output, target_rank, op_name, CPU_DEVICE_ID,
+        [handle, tensor, output, cpu_buffer_output, device, self_weight,
+         pair_weight, op_name, tid,
+         timeline_ptr](const Status& status) mutable {
+          // Will execute in the `device` context.
+          if (status.ok()) {
+            with_device device_guard(device);
+            output.copy_(cpu_buffer_output);
+
+            if (self_weight == 0.5 && pair_weight == 0.5) {
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/torch/mpi_ops.cc:493]: Done

File: scripts/mpi_passive_recv.cc (hunk 1 above)
Comment (ybc1991) [reply scripts/mpi_passive_recv.cc:19]: I don't know. That is what listed in nccl example.

**Pull request merged**